File: src/main/java/no/novari/flyt/archive/gateway/dispatch/file/FilesDispatchService.java
package no.novari.flyt.archive.gateway.dispatch.file;

import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.archive.gateway.dispatch.DispatchStatus;
import no.novari.flyt.archive.gateway.dispatch.file.result.FileDispatchResult;
import no.novari.flyt.archive.gateway.dispatch.file.result.FilesDispatchResult;
import no.novari.flyt.archive.gateway.dispatch.model.instance.DokumentobjektDto;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Collection;
import java.util.Map;

import static java.util.stream.Collectors.toMap;

@Service
@Slf4j
public class FilesDispatchService {

    private final FileDispatchService fileDispatchService;


    public FilesDispatchService(
            FileDispatchService fileDispatchService
    ) {
        this.fileDispatchService = fileDispatchService;
    }

    public Mono<FilesDispatchResult> dispatch(Collection<DokumentobjektDto> dokumentobjektDtos) {
        if (dokumentobjektDtos.isEmpty()) {
            log.info("No files to dispatch");
            return Mono.just(FilesDispatchResult.accepted(Map.of()));
        }
        log.info("Dispatching files");
        return Flux.fromStream(dokumentobjektDtos.stream())
                .concatMap(fileDispatchService::dispatch)
                .takeUntil(fileDispatchResult -> fileDispatchResult.getStatus() != DispatchStatus.ACCEPTED)
                .collectList()
                .map(fileDispatchResults -> {

                    FileDispatchResult lastResult = fileDispatchResults.getLast();
                    DispatchStatus lastStatus = lastResult.getStatus();

                    return switch (lastStatus) {
                        case ACCEPTED -> FilesDispatchResult.accepted(
                                fileDispatchResults
                                        .stream()
                                        .collect(toMap(
                                                FileDispatchResult::getFileId,
                                                FileDispatchResult::getArchiveFileLink
                                        ))
                        );
                        case DECLINED -> FilesDispatchResult.declined(
                                lastResult.getErrorMessage()
                        );
                        case FAILED -> FilesDispatchResult.failed();

                    };

                }).doOnNext(result -> log.info("Dispatch result={}", result.toString()));
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/file/FilesDispatchService.kt
package no.novari.flyt.archive.gateway.dispatch.file

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.flyt.archive.gateway.dispatch.DispatchStatus
import no.novari.flyt.archive.gateway.dispatch.file.result.FileDispatchResult
import no.novari.flyt.archive.gateway.dispatch.file.result.FilesDispatchResult
import no.novari.flyt.archive.gateway.dispatch.model.instance.DokumentobjektDto
import org.springframework.stereotype.Service

@Service
class FilesDispatchService(
    private val fileDispatchService: FileDispatchService,
) {
    fun dispatch(dokumentobjektDtos: Collection<DokumentobjektDto>): FilesDispatchResult {
        if (dokumentobjektDtos.isEmpty()) {
            log.info { "No files to dispatch" }
            return FilesDispatchResult.accepted(emptyMap())
        }

        log.info { "Dispatching files" }
        val fileDispatchResults = mutableListOf<FileDispatchResult>()
        for (dokumentobjektDto in dokumentobjektDtos) {
            val result = fileDispatchService.dispatch(dokumentobjektDto)
            fileDispatchResults += result
            if (result.status != DispatchStatus.ACCEPTED) {
                break
            }
        }

        val lastResult = fileDispatchResults.last()
        val result =
            when (lastResult.status) {
                DispatchStatus.ACCEPTED -> {
                    FilesDispatchResult.accepted(
                        fileDispatchResults
                            .mapNotNull { fileDispatchResult ->
                                val fileId = fileDispatchResult.fileId
                                val archiveFileLink = fileDispatchResult.archiveFileLink
                                if (fileId != null && archiveFileLink != null) {
                                    fileId to archiveFileLink
                                } else {
                                    null
                                }
                            }.toMap(),
                    )
                }

                DispatchStatus.DECLINED -> {
                    FilesDispatchResult.declined(lastResult.errorMessage.orEmpty())
                }

                DispatchStatus.FAILED -> {
                    FilesDispatchResult.failed()
                }
            }
        log.atInfo {
            message = "Dispatch result={}"
            arguments = arrayOf(result)
        }
        return result
    }

    companion object {
        private val log = KotlinLogging.logger {}
    }
}
